cider-dap/src/adapter.rs

```diff
@@ -1,12 +1,74 @@
+use dap::types::{Breakpoint, Source, SourceBreakpoint, Thread};
 use std::fs::File;
 pub struct MyAdapter {
     #[allow(dead_code)]
     file: File,
-    // Other fields of the struct
+    breakpoints: Vec<(Source, i64)>,
+    break_count: Counter,
+    threads: Vec<Thread>,
 }
 
 impl MyAdapter {
     pub fn new(file: File) -> Self {
-        MyAdapter { file }
+        MyAdapter {
+            file,
+            breakpoints: Vec::new(),
+            break_count: Counter::new(),
+            threads: Vec::new(),
+        }
+    }
+
+    //Set breakpoints for adapter
+    pub fn set_breakpoint(
+        &mut self,
+        path: Source,
+        source: &Vec<SourceBreakpoint>,
+    ) -> Vec<Breakpoint> {
+        //Keep all the new breakpoints made
+        let mut out_vec: Vec<Breakpoint> = vec![];
+
+        //Loop over all breakpoints
+        for source_point in source {
+            self.breakpoints.push((path.clone(), source_point.line));
+            //Create new Breakpoint instance
+            let breakpoint = Breakpoint {
+                id: self.break_count.increment().into(),
+                verified: true,
+                message: None,
+                source: Some(path.clone()),
+                line: None,
+                column: None,
+                end_line: None,
+                end_column: None,
+                instruction_reference: None,
+                offset: None,
+            };
+
+            out_vec.push(breakpoint);
+        }
+
+        out_vec
+    }
+
+    //Return threads
+    pub fn get_threads(&self) -> Vec<Thread> {
```

Comment:
rename this something like `clone_threads` since I would normally expect a get method to return a reference, but we need the owned vec for the return message (which is silly but hey that's a problem for another day)